Pinta.Effects/Effects/ForwardErrorDiffusionDitheringEffect.cs

```diff
@@ -0,0 +1,325 @@
+	}
+
+	public override void Render (ImageSurface src, ImageSurface dest, ReadOnlySpan<RectangleI> rois)
+	{
+		var src_data = src.GetReadOnlyPixelData ();
+		var src_copy = new ColorBgra[src_data.Length];
+		src_data.CopyTo (src_copy);
+		var dst_data = dest.GetPixelData ();
+		foreach (var rect in rois) {
+			for (int y = 0; y < rect.Height - Data.DiffusionMatrix.RowsBelow; y++) {
+				for (int x = Data.DiffusionMatrix.ColumnsToLeft; x < rect.Width - Data.DiffusionMatrix.ColumnsToRight; x++) {
+					var currentIndex = y * src.Width + x;
+					var originalPixel = src_copy[currentIndex];
+					var closestColor = FindClosestPaletteColor (originalPixel);
+					dst_data[currentIndex] = closestColor;
+					int errorRed = originalPixel.R - closestColor.R;
+					int errorGreen = originalPixel.G - closestColor.G;
+					int errorBlue = originalPixel.B - closestColor.B;
+					DistributeError (src_copy, x, y, errorRed, errorGreen, errorBlue, src.Width, src.Height);
+				}
+			}
+		}
+	}
+
+	private void DistributeError (Span<ColorBgra> original, int x, int y, int errorRed, int errorGreen, int errorBlue, int sourceWidth, int sourceHeight)
+	{
+		for (int r = 0; r < Data.DiffusionMatrix.Rows; r++) {
+			for (int c = 0; c < Data.DiffusionMatrix.Columns; c++) {
+				if (Data.DiffusionMatrix[r, c] is not WeightElement weight)
+					continue;
+				var this_y = y + r;
+				var this_x = x + c - Data.DiffusionMatrix.ColumnsToLeft;
+				if (this_x >= sourceWidth)
+					continue;
+				if (this_y >= sourceHeight)
+					continue;
+				int idx = (this_y * sourceWidth) + this_x;
+				double factor = ((double) weight.Weight) / Data.DiffusionMatrix.TotalWeight;
+				original[idx] = AddError (original[idx], factor, errorRed, errorGreen, errorBlue);
+			}
+		}
+	}
+
+	private static ColorBgra AddError (ColorBgra color, double factor, int errorRed, int errorGreen, int errorBlue)
+	{
+		// This function will add the error to the color based on the provided factor
+		byte newR = Utility.ClampToByte (color.R + (int) (factor * errorRed));
+		byte newG = Utility.ClampToByte (color.G + (int) (factor * errorGreen));
+		byte newB = Utility.ClampToByte (color.B + (int) (factor * errorBlue));
+		return ColorBgra.FromBgra (newB, newG, newR, 255);
+	}
+
+	private ColorBgra FindClosestPaletteColor (ColorBgra original)
+	{
+		double minDistance = double.MaxValue;
+		ColorBgra closestColor = ColorBgra.FromBgra (0, 0, 0, 1);
+		foreach (var paletteColor in Data.Palette) {
+			double distance = CalculateDistance (original, paletteColor);
+			if (distance < minDistance) {
+				minDistance = distance;
+				closestColor = paletteColor;
+			}
+		}
+		return closestColor;
+	}
+
+	private static double CalculateDistance (ColorBgra color1, ColorBgra color2)
+	{
+		int deltaR = color1.R - color2.R;
+		int deltaG = color1.G - color2.G;
+		int deltaB = color1.B - color2.B;
+
+		// Euclidean distance
+		return Math.Sqrt (deltaR * deltaR + deltaG * deltaG + deltaB * deltaB);
```

Comment:
I think working in terms of squared distance would be fine here and let you avoid the square root and keep things in integers?